### PR TITLE
Implement TimeExchange PACT and extend Exchange operator

### PR DESCRIPTION
**Pull request by antiguru**
Implement exchanging by time and data vs. by data only. This is useful to have permuting worker assignments for skewed data.

Signed-off-by: Moritz Hoffmann <moritz.hoffmann@inf.ethz.ch>

### CODE REVIEW

File: src/dataflow/channels/pact.rs [hunk 1]
```diff
@@ -47,6 +47,27 @@ impl<T: Eq+Data+Abomonation, D: Data+Abomonation, F: Fn(&D)->u64+'static> Parall
     fn connect<A: Allocate>(self, allocator: &mut A, identifier: usize) -> (Box<Push<(T, Content<D>)>>, Box<Pull<(T, Content<D>)>>) {
         let (senders, receiver) = allocator.allocate::<Message<T, D>>();
         let senders = senders.into_iter().enumerate().map(|(i,x)| Pusher::new(x, allocator.index(), i, identifier)).collect::<Vec<_>>();
+        (Box::new(ExchangeObserver::new(senders, move |_, d| (self.hash_func)(d))), Box::new(Puller::new(receiver, allocator.index(), identifier)))
+    }
+}
+
+/// An exchange between multiple observers
+pub struct TimeExchange<D, T, F: Fn(&T, &D)->u64+'static> { hash_func: F, phantom: PhantomData<D>, phantom_t: PhantomData<T>, }
```
Comment (frankmcsherry) [inline src/dataflow/channels/pact.rs]: If you like (it's minor) you can totally use `PhantomData<(D, T)>`. I think that is how folks do it when they have piles of phantom types.

**Comment by frankmcsherry** [comment]
This looks good. I have a few minor doc-requests, which if you could do <3 but I can totally do too:

1. `TimeExchange` description is the same as `Exchange`. 
2. `TimeExchange::new()` says it allocates an `Exchange`.
3. The `exchange_ts` method doesn't describe what it guarantees (which I think is "all records whose time and data evaluate to the same `route` arrive at the same worker" or something like that).

I noticed the build broken due to `RadixSort`. I think I pushed some fixes for that, but perhaps I've screwed that up too. I'll double check!

**Comment by frankmcsherry** [comment]
Ok, I just shoveled a commit at you (at master) that fixes `examples/bfs.rs`. It should be an automatic merge for you, and just fixes an issue of my doing that was causing Travis to complain.

**Comment by antiguru** [comment]
Thanks for fixing the example! I updated the documentation to reflect the differences between `Exchange` and `TimeExchange`.

**Comment by frankmcsherry** [comment]
Looks great, thank you! Once travis comes back, we'll be a-merging. Thanks very much for the contribution. :D

**Pull request merged**